instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/RouteRefreshOnDemandTest.kt

```diff
@@ -174,6 +178,193 @@ class RouteRefreshOnDemandTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::
+            }
+
+            mapboxNavigation.routeRefreshController.requestImmediateRouteRefresh()
+
+            val refresh = withTimeout(3000) {
+                mapboxNavigation.routesUpdates().first { it.reason == RoutesExtra.ROUTES_UPDATE_REASON_REFRESH }
+            }
+            assertEquals(reroute.navigationRoutes.map { it.id }, refresh.navigationRoutes.map { it.id })
+
+            assertEquals(
+                listOf(
+                    RouteRefreshExtra.REFRESH_STATE_STARTED,
+                    RouteRefreshExtra.REFRESH_STATE_FINISHED_SUCCESS,
+                ),
+                observer.getStatesSnapshot()
+            )
+        }
+    }
+
+    @Test
+    fun routes_are_updated_while_refresh_on_demand_request_is_in_progress() = sdkTest {
+        createMapboxNavigation(createRouteRefreshOptionsWithInvalidInterval(100_000))
+        mockWebServerRule.requestHandlers.add(
+            MockDirectionsRequestHandler(
+                "driving-traffic",
+                readRawFileText(activity, R.raw.route_response_route_refresh),
+                twoCoordinates
+            )
+        )
+        val refreshHandler = MockDirectionsRefreshHandler(
+            "route_response_route_refresh",
+            readRawFileText(activity, R.raw.route_response_route_refresh_annotations),
+        )
+        refreshHandler.jsonResponseModifier = DelayedResponseModifier(3000, DynamicResponseModifier())
+        mockWebServerRule.requestHandlers.add(refreshHandler)
+        val observer = TestObserver()
+
+        mapboxNavigation.routeRefreshController.registerRouteRefreshStateObserver(observer)
+        mapboxNavigation.startTripSession()
+        val routeOptions = generateRouteOptions(twoCoordinates)
+        val requestedRoutes = mapboxNavigation.requestRoutes(routeOptions)
+            .getSuccessfulResultOrThrowException()
+            .routes
+        val initialRoutes = requestedRoutes.take(1)
+        mapboxNavigation.setNavigationRoutesAndWaitForUpdate(initialRoutes)
+
+        val refreshes = mutableListOf<RoutesUpdatedResult>()
+        mapboxNavigation.registerRoutesObserver {
+            if (it.reason == RoutesExtra.ROUTES_UPDATE_REASON_REFRESH) {
+                refreshes.add(it)
+            }
+        }
+
+        mapboxNavigation.routeRefreshController.requestImmediateRouteRefresh()
+
+        delay(1000)
+        mapboxNavigation.setNavigationRoutesAndWaitForAlternativesUpdate(requestedRoutes)
+
+        delay(3000)
+        assertEquals(0, refreshes.size)
+
+        mapboxNavigation.routeRefreshController.requestImmediateRouteRefresh()
+        val refreshedRoutes = withTimeout(7000) {// 2 routes with delay 3s each
+            mapboxNavigation.routesUpdates().first {
+                it.reason == RoutesExtra.ROUTES_UPDATE_REASON_REFRESH
+            }.navigationRoutes
+        }
+        assertEquals(requestedRoutes.map { it.id }, refreshedRoutes.map { it.id })
```

Comment:
Do we still need response delay here? 
```suggestion

        refreshHandler.jsonResponseModifier = { it }
        mapboxNavigation.routeRefreshController.requestImmediateRouteRefresh()
        val refreshedRoutes = withTimeout(2000) {
            mapboxNavigation.routesUpdates().first {
                it.reason == RoutesExtra.ROUTES_UPDATE_REASON_REFRESH
            }.navigationRoutes
        }
        assertEquals(requestedRoutes.map { it.id }, refreshedRoutes.map { it.id })
```